app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,80 @@
+module Users
+  class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+    def google_oauth2
+      if !CheckOmniauth.verified?(auth)
+        when_social_profile_not_verified
+      else
+        handle_callback
+      end
+    end
+
+    def facebook
+      if !CheckOmniauth.verified?(auth)
+        when_social_profile_not_verified
+      else
+        handle_callback
+      end
+    end
+
+    private
+
+    def handle_callback
```

Comment:
Well, I have quite concerning solution for this huge controller.

It feels like we can split this method into some services, which will perform necessary actions, and return two things:
1) text for notice
2) URL for redirect

Those three services will respond for two main cases of handling callbacks:

1) Service to determine, which service to call (it feels like we need an organizer + interactors structure or something like that)

2) When we have signed in user (add/remove social accounts from profile)
3) When we authenticate from scratch (find existing user by oauth response)
4) And if we have none - create new user.

Quite concerning solution, just because it's not really clear, but it feels like it can really simplify this controller.